Userland/Applications/FTPServer/FTPServerClient.cpp

```diff
@@ -0,0 +1,1165 @@
+    : m_id(id)
+    , m_socket(move(socket))
+    , m_allow_anonymous(allow_anonymous)
+    , m_working_dir(root_directory)
+{
+    m_socket->on_ready_to_read = [this] { drain_socket(); };
+}
+
+static String format_filesystem_object(const String& name, const struct stat& st)
+{
+    StringBuilder builder;
+
+    if (S_ISDIR(st.st_mode))
+        builder.append("d");
+    else if (S_ISLNK(st.st_mode))
+        builder.append("l");
+    else if (S_ISBLK(st.st_mode))
+        builder.append("b");
+    else if (S_ISCHR(st.st_mode))
+        builder.append("c");
+    else if (S_ISFIFO(st.st_mode))
+        builder.append("f");
+    else if (S_ISSOCK(st.st_mode))
+        builder.append("s");
+    else if (S_ISREG(st.st_mode))
+        builder.append("-");
+    else
+        builder.append("?");
+
+    builder.appendff("{}{}{}{}{}{}{}{}",
+        st.st_mode & S_IRUSR ? 'r' : '-',
+        st.st_mode & S_IWUSR ? 'w' : '-',
+        st.st_mode & S_ISUID ? 's' : (st.st_mode & S_IXUSR ? 'x' : '-'),
+        st.st_mode & S_IRGRP ? 'r' : '-',
+        st.st_mode & S_IWGRP ? 'w' : '-',
+        st.st_mode & S_ISGID ? 's' : (st.st_mode & S_IXGRP ? 'x' : '-'),
+        st.st_mode & S_IROTH ? 'r' : '-',
+        st.st_mode & S_IWOTH ? 'w' : '-');
+
+    if (st.st_mode & S_ISVTX)
+        builder.append("t");
+    else
+        builder.appendff("{}", st.st_mode & S_IXOTH ? 'x' : '-');
+
+    builder.appendff(" {}", st.st_nlink);
+
+    builder.appendff(" {}", st.st_uid);
+
+    builder.appendff(" {}", st.st_gid);
+
+    if (S_ISCHR(st.st_mode) || S_ISBLK(st.st_mode)) {
+        builder.appendff("  {},{} ", major(st.st_rdev), minor(st.st_rdev));
+    } else {
+        builder.appendff(" {} ", (uint64_t)st.st_size);
+    }
+
+    builder.appendff("  {}  ", Core::DateTime::from_timestamp(st.st_mtime).to_string("%h %d  %Y"));
+
+    builder.append(name);
+    builder.append("\r\n");
+
+    return builder.to_string();
+}
+
+void FTPServerClient::drain_socket()
+{
+    NonnullRefPtr<FTPServerClient> protect(*this);
+
+    while (m_socket && m_socket->can_read()) {
+        if (!m_socket || m_socket->eof()) {
+            quit();
+            break;
+        }
+
+        auto buf = m_socket->read(4 * KiB);
+
+        if (buf.is_empty()) {
+            quit();
+            break;
+        }
+
+        // remove \r\n
+        String data = String(buf, AK::NoChomp);
+        data = data.substring(0, data.length() - 2);
+
+        dbgln_if(FTP_DEBUG, "Received: {}", data);
+
+        if (on_receive_command)
+            on_receive_command(this, data);
+
+        handle_command(data);
+    }
+
+    if (!m_socket || m_socket->eof()) {
+        quit();
+        return;
+    }
+}
+
+void FTPServerClient::handle_command(String input)
+{
+    Vector<String> parts = input.split(' ');
+    String command = parts.take_first();
+
+    if (command.equals_ignoring_case("AUTH")) {
```

Comment:
instead of doing a bunch of if statements could this be done with an `AK::HashTable` ?